trace-resources/src/Cardano/Logging/Resources/Darwin.hsc

```diff
@@ -0,0 +1,114 @@
+import qualified GHC.Stats as GhcStats
+import           System.Posix.Process (getProcessID)
+import           System.Posix.Types (ProcessID)
+import           Cardano.Logging.Resources.Types
+
+{- type aliases -}
+type MACH_VM_SIZE_T = Word64
+data TIME_VALUE_T = TIME_VALUE_T Word64 Word64
+
+{- memory information -}
+
+{- mach/task_info.h
+struct time_value {
+	integer_t seconds;
+	integer_t microseconds;
+};
+struct mach_task_basic_info {
+	mach_vm_size_t  virtual_size;       /* virtual memory size (bytes) */
+	mach_vm_size_t  resident_size;      /* resident memory size (bytes) */
+	mach_vm_size_t  resident_size_max;  /* maximum resident memory size (bytes) */
+	time_value_t    user_time;          /* total user run time for
+	                                     *  terminated threads */
+	time_value_t    system_time;        /* total system run time for
+	                                     *  terminated threads */
+	policy_t        policy;             /* default policy for new threads */
+	integer_t       suspend_count;      /* suspend count for task */
+}; -}
+
+data MachTaskBasicInfo = MachTaskBasicInfo
+  { _virtual_size :: !MACH_VM_SIZE_T
+  , _resident_size :: !MACH_VM_SIZE_T
+  , _resident_size_max :: !MACH_VM_SIZE_T
+  , _user_time :: !TIME_VALUE_T
+  , _system_time :: !TIME_VALUE_T
+  , _policy :: !Word64
+  , _suspend_count :: !Word64
+  }
+
+instance Storable TIME_VALUE_T where
+  alignment _ = #const offsetof(struct {char x__; struct time_value (y__); }, y__)
+  sizeOf _    = #size struct time_value
+  peek ptr    = TIME_VALUE_T
+                <$> (#peek struct time_value, seconds) ptr
+                <*> (#peek struct time_value, microseconds) ptr
+  poke _ _    = pure ()
+
+instance Storable MachTaskBasicInfo where
+  alignment _ = #const offsetof(struct {char x__; struct mach_task_basic_info (y__); }, y__)
+  sizeOf _    = #size struct mach_task_basic_info
+  peek ptr    = MachTaskBasicInfo
+                <$> (#peek struct mach_task_basic_info, virtual_size) ptr
+                <*> (#peek struct mach_task_basic_info, resident_size) ptr
+                <*> (#peek struct mach_task_basic_info, resident_size_max) ptr
+                <*> (#peek struct mach_task_basic_info, user_time) ptr
+                <*> (#peek struct mach_task_basic_info, system_time) ptr
+                <*> (#peek struct mach_task_basic_info, policy) ptr
+                <*> (#peek struct mach_task_basic_info, suspend_count) ptr
+  poke _ _    = pure ()
+
+foreign import ccall unsafe c_get_process_memory_info :: Ptr MachTaskBasicInfo -> CInt -> IO CInt
+
+
+getMemoryInfo :: ProcessID -> IO MachTaskBasicInfo
+getMemoryInfo pid =
+  allocaBytes 128 $ \ptr -> do
+    res <- c_get_process_memory_info ptr (fromIntegral pid)
+    if res <= 0
+      then do
+        putStrLn $ "c_get_process_memory_info: failure returned: " ++ (show res)
+        return $ MachTaskBasicInfo 0 0 0 (TIME_VALUE_T 0 0) (TIME_VALUE_T 0 0) 0 0
+      else
+        peek ptr
```

Comment:
One could also use `throwIfNeg`, though one needs to account that it checks if the value is `< 0` rather than `<= 0`  (maybe this is not even needed).
```suggestion
    rthrowIfNeg_ (\res -> "c_get_process_memory_info: failure returned: " ++ show (pred res))
                  (succ <$> c_get_process_memory_info ptr (fromIntegral pid))
    peek ptr
 
```

Reply:
As I have no Mac it's a big pain to do changes here. I have to wait for the mac and windows builds on github to see if it compiles at all. So someone with a mac could try this?

Reply:
I don't have a mac, but this should be the same on linux and macos.  You could try it out first in `Cardano.Logging.Resources.Linux`.

Reply:
Do you use `nix`? Then there's a command to cross compile it on linux for macos.  @deepfire might know the command.  I'm sometimes cross compiling from linux to windows using `haskell.nix` which is available in our repos.

Reply:
> You could try it out first in `Cardano.Logging.Resources.Linux`.
Linux doesn't have a C part here. 
> Do you use `nix`? Then there's a command to cross compile it on linux for macos. @deepfire might know the command. 
I already spend some time on this without success, and of course asked Serge. Seems these tools are currently not working.
But I will work on making changes to the c code as suggested.

> One could also use `throwIfNeg`, though one needs to account that it checks if the value is `< 0` rather than `<= 0` (maybe this is not even needed).
I don't even know where this function comes from, and don't see the advantage here anyway

Reply:
I think we've addressed this with the suggested code change, @coot

Reply:
Done

Reply:
I still see `res <= 0` above.  The [throwIfNeg_](https://hackage.haskell.org/package/base-4.15.0.0/docs/Foreign-Marshal-Error.html#v:throwIfNeg_) is provided by `base` package.

Reply:
> I don't even know where this function comes from, and don't see the advantage here anyway

It's a common C practice to use `< 0` for errors and `0` for no-error, keeping this, will be easier for an (independent) reviewer.

Reply:
Changed this for Windows as well. Funny enough the underlying Windows call returns 1 for success and 0 for failure :)

Reply:
yes I think it's more common for Windows to use `0` for failure, though I am not sure how consistent they are.